### PR TITLE
Allow configuration of PlantUML includes

### PR DESCRIPTION
**Pull request by stephenc**
We generate images as part of our build chain and so we trust the URLs and other libraries that we want to use as includes. The current includes blocking is heavy handed. This should allow import path includes as well as enabling a whitelist of regexes for anyone who wants to allow including only a subset of URLs that they actually trust.

Note: I would advise repeating my analysis of PlantUML's `ImportedFiles` to verify that for regular file imports it only permits the inclusion of files that are actually on the include path (my analysis showed that it is a bit too strict as you cannot have `!include foo/bar.puml` because the `getSystemFolder()` will then return `$DIR_FROM_SEARCH_PATH/foo` which is not explicitly one of the files on the search path... but that seems sufficiently strict from my PoV to allow any paths that do not contain `..` style navigation and are not URLs

### CODE REVIEW

**Comment by Mogztter** [comment]
Wow thanks for this great contribution!
This is a popular request and we will probably extend it (in the future) to Vega and Vega-Lite which are using a similar feature to include data from URL.

Overall I like the general concept but I would like to keep the `SECURE` safe mode as is and introduce a new level `SAFE` where the include path and the include whitelist will be taken into consideration.

Basically, `SECURE` level does not trust the underlying library.
It's not really related to this pull request but, just for reference, on the public instance of Kroki I would like to deny access to all/most filesystem operations using Docker (ie. remove Linux kernel capabilities).

File: server/README.adoc [hunk 1]
```diff
@@ -49,6 +53,24 @@ To keep the Docker image as small as possible, the following libraries are provi
 
 When one of the above library is requested, Kroki server will delegate the call.
 
+==== PlantUML include files
+
+As PlantUML syntax includes pre-processor directives that have the potential to access external resources, serving live content from Kroki server is only recommended when using the dfeault settings of `KROKI_SAFE_MODE=SECURE` and `KROKI_PLANTUML_ALLOW_INCLUDE=false`.
```
Comment (Mogztter) [inline server/README.adoc:58]:
```suggestion
As PlantUML syntax includes pre-processor directives that have the potential to access external resources, serving live content from Kroki server is only recommended when using the default settings of `KROKI_SAFE_MODE=SECURE` and `KROKI_PLANTUML_ALLOW_INCLUDE=false`.
```

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 2]
```diff
@@ -24,25 +24,52 @@
...
+   *   <li>{@code !include /absolute/file/path} includes a file from the filesystem</li>
+   *   <li>{@code !include search-path-file} includes a file on the "plantuml.include.path"</li>
+   *   <li>{@code !include search-path-file!2} includes the 3rd document in a file on the "plantuml.include.path"</li>
+   *   <li>{@code !includesub search-path-file!SUBID} includes the SUBID subpart of a file on the "plantuml.include.path"</li>
+   *   <li>{@code !includeurl http://some/url} deprecated include of an external resource by URL</li>
+   * </ul>
+   *
+   * @see <a href="https://plantuml.com/preprocessing">PlantUML Preprocessing</a>
+   */
+  private static final Pattern INCLUDE_RX = Pattern.compile("^\\s*!include(?:url|sub)?\\s+(.*)(?:!.*)?(?:#.*)?");
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:66]: Much needed documentation, thanks :+1: 
What is the last part of the regular expression with the `#` symbol? Is that when a URL contains a fragment?

File: server/README.adoc [hunk 3]
```diff
@@ -28,6 +28,10 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
 KROKI_NOMNOML_BIN_PATH:: Path to the `nomnoml` binary
 KROKI_BLOCKDIAG_HOST:: The host of the blockdiag server
 KROKI_BLOCKDIAG_PORT:: The port of the blockdiag server
+KROKI_SAFE_MODE:: Either `SECURE` (default) or `UNSAFE`. Determines whether to sanitize requests for formats that allow including external content
+KROKI_PLANTUML_INCLUDE_PATH:: The include path to set for PlantUML.
+KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list or regexes for valid includes.
```
Comment (Mogztter) [inline server/README.adoc:33]: I guess that passing a list of regexes as environment variable is not practical but using a file means that you need to extend the docker image or bind mount a directory with this file right?
Is there an alternative solution?

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 4]
```diff
@@ -65,8 +92,26 @@ public String decode(String encoded) throws DecodeException {
...
+    }
+  }
+
+  private static List<Pattern> parseIncludeWhitelist(String filename) {
+    if (filename != null && new File(filename).isFile()) {
+      try {
+        return Files.lines(Paths.get(filename)).filter(s -> !s.isEmpty()).map(Pattern::compile)
+          .collect(Collectors.toList());
+      } catch (IOException e) {
+        // ignore
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:111]: I think we should log the error with the context, so the user knows what is going on:

```
logger.warn("Unable to read the PlantUML whitelist file: xyz. The include whitelist will be empty.", e)
```

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 5]
```diff
@@ -65,8 +92,26 @@ public String decode(String encoded) throws DecodeException {
...
+    OptionFlags.ALLOW_INCLUDE = config.getBoolean("KROKI_PLANTUML_ALLOW_INCLUDE", false);
+    String plantUmlIncludePath = config.getString("KROKI_PLANTUML_INCLUDE_PATH");
+    if (plantUmlIncludePath != null) {
+      // NOTE: relying on classloading sequencing to ensure this gets set before ImportedFiles class is loaded by JVM
+      System.setProperty("plantuml.include.path", plantUmlIncludePath);
+    }
+  }
+
+  private static List<Pattern> parseIncludeWhitelist(String filename) {
+    if (filename != null && new File(filename).isFile()) {
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:106]: If the filename is not a file or is not readable I also think we should add a log.

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 6]
```diff
@@ -177,6 +226,43 @@ private static void ignoreInclude(String line, StringBuilder sb) {
...
+        //   a fixed set guarded by `/stdlib/${name}-abx.repx` resources on the classpath, in any case we
+        //   ignore stdlib import validation here
+        // * Absolute imports could refer to any file, so we exclude those, though FileWithSuffix#fileOk()
+        //   should restrict accessible files to only those "approved"
+        // * We only auto-permit "search path" includes that do not attempt search path escaping with `..`
+        // * The import or importsub that we auto-permit will result in a call to ImportedFiles.getFile(...)
+        // * That will iterate the "plantuml.include.path" (note PlantUML is not well designed for embedding
+        //   on this property as the ImportedFiles#INCLUDE_PATH is initialized once on classloading)
+        // * At this point, if the INCLUDE_PATH included say `/example` and the include was `foo/bar.puml`
+        //   and there is a file `/example/foo/bar/puml` then that would stop the search with an `AFileRegular`
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:253]: you meant `/example/foo/bar.puml`?

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 7]
```diff
@@ -177,6 +226,43 @@ private static void ignoreInclude(String line, StringBuilder sb) {
...
+        //   instance. HOWEVER, the return value of ImportedFiles#getFile is guarded by ImportedFiles#isAllowed
+        // * ImportedFiles#isAllowed will only permit AFile instances with the AFile#getSystemFolder() being
+        //   contained in ImportedFiles#INCLUDE_PATH so as AFileRegular#getSystemFolder() always returned the
+        //   parent folder of the file and include of `foo/bar.puml` found on an include path of `/example`
+        //   resolving to `/example/foo/bar.puml` will have a system folder of `/example/foo` which is not
+        //   one of the folders whitelisted in the include path and thus the include will be resolved as not-found
+        //
+        // Effectively only imports that are immediate children of the folders listed in "plantuml.include.path"
+        // are eligible for inlcude (unless OptionFlags.ALLOW_INCLUDE has been set to true)
+        sb.append(line).append("\n");
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:263]: I really appreciate the time and effort you put into this clear and precise analysis of the PlantUML include checking algorithm :100: 

If I understand correctly, even if we do allow path walking, the PlantUML include algorithm should not resolve the include if the path is outside of the include path right? So if my include path is `/example` and my include is `../etc/passwd`, the system folder of the file will be `/etc` and `/etc` is not part of the include path `/example` so the include will be resolved as not-found. Am I correct?
I'm not saying that we should remove the path walking checks (as we want to disallow them as soon as possible) but I want to make sure that I understand how it's working.

File: server/src/main/java/io/kroki/server/service/Plantuml.java (hunk 2 above)
Comment (stephenc) [reply server/src/main/java/io/kroki/server/service/Plantuml.java:66]: that's a comment in plantuml

File: server/README.adoc (hunk 3 above)
Comment (stephenc) [reply server/README.adoc:33]: you could have one regex per env-var and chase up env vars with indexes, e.g. `KROKI_PLANTUML_INCLUDE_WHITELIST_0`, `KROKI_PLANTUML_INCLUDE_WHITELIST_1`, until you stop finding `KROKI_PLANTUML_INCLUDE_WHITELIST_n` for some integer n

File: server/src/main/java/io/kroki/server/service/Plantuml.java (hunk 7 above)
Comment (stephenc) [reply server/src/main/java/io/kroki/server/service/Plantuml.java:263]: correct, but please verify my claim rather than take it at face value. I believe it to be correct, but you are the maintainer, so "trust but verify" especially as this as security implications

File: server/README.adoc [hunk 8]
```diff
@@ -49,6 +71,24 @@ To keep the Docker image as small as possible, the following libraries are provi
...
+
+The recommended pattern for handling includes is to either extend the docker image or bind mount a directory containing the files you wish to include and then set the `KROKI_PLANTUML_INCLUDE_PATH` environment variable.
+
+For example:
+
+[source,bash]
+----
+$ mkdir example-lib
+$ printf "@startuml\nBob->Alice\n@enduml\n" > example-lib/bob.puml
+$ sudo docker run --publish=8000:8000 -v $(pwd)/example-lib:/example-lib -e KROKI_PLANTUML_INCLUDE_PATH=/example-lib yuzutech/kroki:latest
```
Comment (stephenc) [inline server/README.adoc:88]:
```suggestion
$ sudo docker run --publish=8000:8000 -v $(pwd)/example-lib:/example-lib -e KROKI_SAFE_MODE=SAFE -e KROKI_PLANTUML_INCLUDE_PATH=/example-lib yuzutech/kroki:latest
```

**Comment by stephenc** [comment]
Anything else?

**Comment by Mogztter** [comment]
> Anything else?

No it's looking really good! :+1: 
I will probably do a thorough review/testing session this weekend and I will also take time to read the source code of PlantUML. I will let you know :wink:

**Comment by Mogztter** [comment]
@stephenc I'm looking at it right now, I will probably do some small changes and commit them.

File: server/src/main/java/io/kroki/server/service/Plantuml.java [hunk 9]
```diff
@@ -153,30 +229,73 @@ static String withDelimiter(String source) {
...
+      } else if (includeWhitelist.stream().anyMatch(p -> p.matcher(include).matches())) {
```
Comment (Mogztter) [inline server/src/main/java/io/kroki/server/service/Plantuml.java:297]: Not sure which parameter should have the priority between the safe mode and the include whitelist options. I am somewhat leaning toward safe mode. In other words, I think that if safe mode SECURE is defined we should ignore whitelist. Arguably it's a bit restrictive but if it's clearly stated in the documentation that's OK.
Plus, if we do the opposite you cannot have this behavior.

Comment (Mogztter) [reply server/src/main/java/io/kroki/server/service/Plantuml.java]: I went ahead and now safe mode SECURE has the priority over the whitelist.

File: server/README.adoc [hunk 10]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
 KROKI_NOMNOML_BIN_PATH:: Path to the `nomnoml` binary
 KROKI_BLOCKDIAG_HOST:: The host of the blockdiag server
 KROKI_BLOCKDIAG_PORT:: The port of the blockdiag server
+KROKI_SAFE_MODE:: One of:
++
+--
+`SECURE`:: (default) Paranoid sanitization of requests before passing to image server.
```
Comment (Mogztter) [inline server/README.adoc:34]: I have used "diagram library" instead of "image server" in the documentation. I'm not a native speaker so maybe "image server" is more appropriated but in this case, we should replace every occurrences of "diagram library" otherwise it might be confusing.

Maybe we can stick with "diagram library" in this pull request and create another one to change the terminology if needed? What do you think?

File: server/README.adoc [hunk 11]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
...
+--
++
+[NOTE]
+--
+Some of the image servers allow referencing external entities by URL or accessing resources from the filesystem.
+
+For example PlantUML allows the `!import` directive to pull fragments from the filesystem or a remote URL or the standard library.
+
+It is the responsibility of the upstream codebases to ensure that they can be safely used without risk.
+Because Kroki does not perform code review of these servies, our default setting is to be paranoid and block imports unless known safe.
```
Comment (Mogztter) [inline server/README.adoc:46]:
```suggestion
Because Kroki does not perform code review of these services, our default setting is to be paranoid and block imports unless known safe.
```

File: server/README.adoc [hunk 12]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
...
+We encourage anyone running their own Kroki server to review the services security settings and select the security mode appropriate for their use case.
```
Comment (Mogztter) [inline server/README.adoc:47]: Not sure about this one:

```suggestion
We encourage anyone running their own Kroki server to review the services security settings and select the appropriated security mode for their use case.
```

File: server/README.adoc [hunk 13]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
...
+--
+KROKI_PLANTUML_INCLUDE_PATH:: The include path to set for PlantUML.
+KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list or regexes for valid includes.
```
Comment (Mogztter) [inline server/README.adoc:50]:
```suggestion
KROKI_PLANTUML_INCLUDE_WHITELIST:: The name of a file that consists of a list of Java regular expressions for valid includes.
```

File: server/README.adoc (hunk 12 above)
Comment (stephenc) [reply server/README.adoc:47]: As a native English speaker, don’t make this change

Comment (Mogztter) [reply server/README.adoc:47]: Alright :baguette_bread: :fr:

File: server/README.adoc (hunk 10 above)
Comment (Mogztter) [reply server/README.adoc:34]: So stick with "diagram library"?

Comment (Mogztter) [reply server/README.adoc:34]:
```suggestion
`SECURE`:: (default) Paranoid sanitization of requests before passing to the diagram library.
```

File: server/README.adoc [hunk 14]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
 KROKI_NOMNOML_BIN_PATH:: Path to the `nomnoml` binary
 KROKI_BLOCKDIAG_HOST:: The host of the blockdiag server
 KROKI_BLOCKDIAG_PORT:: The port of the blockdiag server
+KROKI_SAFE_MODE:: One of:
++
+--
+`SECURE`:: (default) Paranoid sanitization of requests before passing to image server.
+`SAFE`:: Assume the image servers secure mode request sanitization is sufficient
```
Comment (Mogztter) [inline server/README.adoc:35]:
```suggestion
`SAFE`:: Assume the diagram libraries secure mode request sanitization is sufficient
```

File: server/README.adoc [hunk 15]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
...
+`UNSAFE`:: Run the image servers without request sanitization.
```
Comment (Mogztter) [inline server/README.adoc:36]:
```suggestion
`UNSAFE`:: Run the diagram libraries without request sanitization.
```

File: server/README.adoc [hunk 16]
```diff
@@ -28,6 +28,28 @@ KROKI_DOT_BIN_PATH:: Path to the `dot` binary (provided by GraphViz)
...
+--
++
+[NOTE]
+--
+Some of the image servers allow referencing external entities by URL or accessing resources from the filesystem.
```
Comment (Mogztter) [inline server/README.adoc:41]:
```suggestion
Some of the diagram libraries allow referencing external entities by URL or accessing resources from the filesystem.
```

**Comment by Mogztter** [comment]
@stephenc I think it's ready to go. I made a few changes, I will let you review them before merging.

**Comment by Mogztter** [comment]
I got a thumb up from @stephenc so let's do it!
A big thank you Stephen for your work :medal_sports:
